### PR TITLE
feature `ForEachIdx`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- add index domain decription class `IdxConfig`
- add for each functor to iterate over a index domain

`ForEachIdx` is the base algorithm to unroll loops for the upcoming lock step programming model for each kernel. This allows to declutch the *supercell* size and the number of worker (threads) to process a supercell or the particles within a *supercell*.

## Example
- old implementation only usable for functor with to parameter
```C++
for (int i = threadId; i < math::CT::volume<FullSuperCellSize>::type::value; i += maxThreads)
{
    const DataSpace<Dim> pos(DataSpaceOperations<Dim>::template map<FullSuperCellSize > (i) - OffsetOrigin::toRT());
    f(p1(pos), p2(pos));
}
```
- new implementation, usable for arbitrary number of functor parameter
```C++
mappings::threads::ForEachIdx<
    mappings::threads::IdxConfig<
        math::CT::volume<FullSuperCellSize>::type::value,
        workerSize
    >
>{ m_workerIdx }(
    [&]( uint32_t const linearIdx, uint32_t const )
    {
        DataSpace< dim > const pos(
            DataSpaceOperations< dim >::template map< FullSuperCellSize >( linearIdx ) -
            OffsetOrigin::toRT( )
        );
        functor( args(pos) ... );
    }
);
```

### CODE REVIEW

File: src/libPMacc/include/mappings/threads/IdxConfig.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    struct IdxConfig
+    {
+        /** number of indices within the domain */
+        static constexpr uint32_t domainSize = T_domainSize;
+        /** number of worker (threads) working on @p domainSize */
+        static constexpr uint32_t workerSize = T_workerSize;
+        /** SMID width */
+        static constexpr uint32_t simdSize = T_simdSize;
+
+        /** number of collective iterations needed to address to address all indices */
```
Comment (BenjaminW3) [inline src/libPMacc/include/mappings/threads/IdxConfig.hpp]: `to address to address`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+         * @param workerIdx index of the worker: range [0;workerSize)
+         */
+        HDINLINE
+        ForEachIdx( uint32_t const workerIdx ) : m_workerIdx( workerIdx )
+        {
+        }
+
+        /** execute a functor
+         *
+         * @p functor is called for each index which is mapped to the worker.
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: `@param`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "mappings/threads/IdxConfig.hpp"
+#include "pmacc_types.hpp"
+
+namespace PMacc
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: one more empty line when includes are finished

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp (hunk 2 above)
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: no `.` at the end of a parameter description

Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: maybe an empty line between parameter description and further description

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            typename T_Functor,
+            typename ... T_Args
+        >
+        HDINLINE void
+        operator()(
+            T_Functor const & functor,
+            T_Args && ... args
+        ) const
+        {
+            for( uint32_t i = 0; i < numCollIter; ++i )
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: `0u`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            {
+                const uint32_t beginWorker = i * simdSize;
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: 2x: `uint32_t const`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                const uint32_t beginIdx = beginWorker * workerSize + simdSize * m_workerIdx;
+                if(
+                    outerLoopCondition ||
+                    !innerLoopCondition ||
+                    beginIdx < domainSize
+                )
+                {
+                    for( uint32_t j = 0; j < simdSize; ++j )
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: `0u`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        >
+        HDINLINE void
+        operator()(
+            T_Functor & functor,
+            T_Args && ... args
+        ) const
+        {
+            for( uint32_t i = 0; i < numCollIter; ++i )
+            {
+                const uint32_t beginWorker = i * simdSize;
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: 2x `uint32_t const`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            typename T_Functor,
+            typename ... T_Args
+        >
+        HDINLINE void
+        operator()(
+            T_Functor & functor,
+            T_Args && ... args
+        ) const
+        {
+            for( uint32_t i = 0; i < numCollIter; ++i )
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: `0u`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            {
+                const uint32_t beginWorker = i * simdSize;
+                const uint32_t beginIdx = beginWorker * workerSize + simdSize * m_workerIdx;
+                if(
+                    outerLoopCondition ||
+                    !innerLoopCondition ||
+                    beginIdx < domainSize
+                )
+                {
+                    for( uint32_t j = 0; j < simdSize; ++j )
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: `0u`

File: src/libPMacc/include/mappings/threads/ForEachIdx.hpp [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                    {
+                        const uint32_t localIdx = beginIdx + j;
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ForEachIdx.hpp]: `uint32_t const`

File: src/libPMacc/include/mappings/threads/IdxConfig.hpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    /** describe a constant index domain
+     *
+     * @tparam T_domainSize number of indices in the domain
+     * @tparam T_workerSize number of worker working on @p T_domainSize
+     * @tparam T_simdSize SMID width
+     */
+    template<
+        uint32_t T_domainSize,
+        uint32_t T_workerSize,
+        uint32_t T_simdSize = 1
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/IdxConfig.hpp]: `1u`

File: src/libPMacc/include/mappings/threads/IdxConfig.hpp [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        /** number of indices within the domain */
+        static constexpr uint32_t domainSize = T_domainSize;
+        /** number of worker (threads) working on @p domainSize */
+        static constexpr uint32_t workerSize = T_workerSize;
+        /** SMID width */
+        static constexpr uint32_t simdSize = T_simdSize;
+
+        /** number of collective iterations needed to address all indices */
+        static constexpr uint32_t numCollIter =
+            (( domainSize + simdSize * workerSize - 1u ) / ( simdSize * workerSize));
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/IdxConfig.hpp]: superfluous outer brackets

**Pull request merged**